[Chong Zhi Lin] iP

czlin98 opened the pull request.

czlin98 committed:
Add Level-0

czlin98 committed:
Add Level-1

czlin98 committed:
Add Level-2

czlin98 committed:
Add Level-3

czlin98 committed:
Add Level-4

czlin98 committed:
Add Level-4

czlin98 committed:
Add A-CodeQuality

Kafcis commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,154 @@
+    private static int addTodo(Task[] tasks, int taskCount, String userCommand, int lineSize) {
+        String todoTask = getTask(userCommand);
+        for (Task task : tasks) {
+
```
removing the space will help with readability

czlin98 replied:
Thanks for the feedback I might have overlooked that part

Bryanbeh1998 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,154 @@
+    }
+
+    private static String getTask(String userCommand) {
+        if (userCommand.contains("/")) {
+            return userCommand.substring(userCommand.indexOf(" ") + 1, userCommand.indexOf("/") - 1);
+        }
+        else {
+            return userCommand.substring(userCommand.indexOf(" ") + 1);
+        }
```
Should not have multiple  return lines in a single method

czlin98 replied:
You're right thanks for the feedback

Bryanbeh1998 reviewed (commented):
Very neat code

Kafcis commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,154 @@
+                setDone(tasks, userCommand);
+            }
+            else {
+                int lineSize = userCommand.length();
```
if else class could use K&R style brackets?

Kafcis commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,154 @@
+    private static int addTodo(Task[] tasks, int taskCount, String userCommand, int lineSize) {
+        String todoTask = getTask(userCommand);
+        for (Task task : tasks) {
+
```
removing space will help with readability (duplicate comment because I need to redo the review)

czlin98 committed:
Add Level-5

czlin98 committed:
Add A-Exceptions

czlin98 committed:
Add A-Packages

czlin98 committed:
Add delete task function
Remove NullDoneIndexException
Resolve typo in showInvalidCommandExceptionMessage

czlin98 committed:
Add A-Collections

czlin98 committed:
Add save and read functions
Remove NullDoneIndexException.java
Add toSave() methods to task classes
Implement collections with ArrayList<Task>
Change order of InitProgram() and showWelcomeMessage()

czlin98 committed:
Merge branch 'branch-Level-6'
Add Level-6

czlin98 committed:
Add Level-7
Resolve merge conflict

czlin98 committed:
Change status icons from tick/cross to yes/no

0WN463 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,374 @@
+
+public class Duke {
+
+    public static final String COMMAND_TODO = "todo";
```
Very nice use of constants

0WN463 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    private static void readFromFile() throws FileNotFoundException {
+        File f = new File(FILE_PATH);
+        Scanner s = new Scanner(f);
```
Single letter variables are reserved for iteration.
Try to use more descriptive variable names

0WN463 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,374 @@
+                tasks.add(new Deadline(taskTypeAndParams[2], taskTypeAndParams[3]));
+                break;
+            case LETTER_EVENT:
+                tasks.add(new Event(taskTypeAndParams[2], taskTypeAndParams[3]));
```
You can consider using temporary variables as labels so that reviewers have a better understanding of how your command string is formatted.

0WN463 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,374 @@
+    private static void executeCommand(String userInput) throws InvalidCommandException{
+        String commandType = getCommandType(userInput);
+        String commandArgs = getCommandArgs(userInput);
+        try {
+            switch (commandType) {
+            case COMMAND_TODO:
+                addTodo(commandArgs);
+                break;
+            case COMMAND_DEADLINE:
+                addDeadline(commandArgs);
+                break;
+            case COMMAND_EVENT:
+                addEvent(commandArgs);
+                break;
+            case COMMAND_DONE:
+                markTaskAsDone(commandArgs);
+                break;
+            case COMMAND_LIST:
+                showList();
+                break;
+            case COMMAND_DELETE:
+                deleteTask(commandArgs);
+                break;
+            case COMMAND_BYE:
+                saveTasks();
+                exitProgram();
+                break;
+            default:
+                throw new InvalidCommandException();
+            }
+        }
+        catch (NullCommandArgsException e) {
+            showNullCommandArgsExceptionMessage(commandType);
+        }
```
Very clean business logic + error handling

0WN463 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,374 @@
+            throw new NullCommandArgsException();
+        }
+        else{
+            String[] deadlineDescriptionAndDate = splitDeadlineDescriptionAndDate(commandArgs);
+            String deadlineDescription = deadlineDescriptionAndDate[0];
+            String deadlineDate = deadlineDescriptionAndDate[1];
```
Good use of temporary variables are labels

0WN463 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    private static void writeToFile(String filePath) throws IOException {
+        FileWriter fw = new FileWriter(filePath);
```
Consider using more descriptive variable names.
Don't be afraid of using longer names as IDE can autocomplete it for you
(At the same time, don't go overboard and use **very** long and cumbersome names)

0WN463 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,26 @@
+    public static final String DATE_POSTFIX = ")";
+    public static final String LETTER_DEADLINE = "D";
+    public static final String DELIMITER = " | ";
+    protected String by;
```
Is there a reason why this field is protected instead of private?

0WN463 reviewed (commented):
Very well done. Code is very readable and maintainable due to good use of constants.

Only minor nitpicks here and there

czlin98 committed:
Add temporary variables to elements of taskTypeAndParams
Change single letter variables to descriptive variables
Change protected fields in the task classes to private

czlin98 committed:
Change private fields in the superclass Task to protected so subclasses can access them

czlin98 committed:
Make the code more OOP and update error handling

czlin98 committed:
Add LocalDate objects to store formatted dates

czlin98 committed:
Add ability to find a task by searching for a keyword and implement streams for listing tasks

czlin98 committed:
Merge pull request #1 from czlin98/branch-Level-8

Add LocalDate objects to store formatted dates

czlin98 committed:
Merge branch 'master' into branch-Level-9

czlin98 committed:
Merge pull request #2 from czlin98/branch-Level-9

Branch level 9

czlin98 committed:
Add javadoc comments

czlin98 committed:
Resolve merge conflicts.

czlin98 committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-JavaDoc

czlin98 committed:
Refactor parser methods and extract messages class.

czlin98 committed:
Add ability to store time as LocalTime objects and add help command.

czlin98 committed:
Set theme jekyll-theme-cayman

czlin98 committed:
Set theme jekyll-theme-slate

czlin98 committed:
Organize packages

czlin98 committed:
Merge remote-tracking branch 'origin/master'

czlin98 committed:
Add user guide, change file path, handle IOOB exception

czlin98 committed:
Fix whitespace in user guide

czlin98 committed:
Add saving data in contents of user guide

czlin98 committed:
Fix some typo in user guide

czlin98 committed:
Revert _config.yml and change some layout choices

czlin98 committed:
Set theme jekyll-theme-cayman

czlin98 committed:
Set theme jekyll-theme-minimal

czlin98 committed:
Remove _config.yml and add duke logo

czlin98 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	docs/_config.yml

czlin98 committed:
Upload product screenshots

czlin98 committed:
Add image links in user guide

czlin98 committed:
Add product download link

czlin98 committed:
Fix table rendering issue in user guide

czlin98 committed:
Add _config.yml to enable emojis

czlin98 committed:
Fix image links in user guide

czlin98 committed:
Fix typo in readme from exit to bye
